microsoft/DevHomeAzureExtension
# Add "My Pull Requests" widget, improvements to widget resilience and loading behavior

## Summary of the pull request

This update is a general resilience and experience improvement with widget data handling, as well as adding a new "My Pull Requests" widget that shows all of a developer's pull requests across all repositories. There is no need to specify which repositories or accounts - all are merged together and shown by pull request date.

## References and relevant issues
Closes #248

## Detailed description of the pull request / Additional comments

* Added resilience to data query since the data tables may be locked in a transaction when a widget tries to query the data. Widgets now make use of DataState and have retry logic and better state display when there is an error either updating or reading data.
* Removed sending of events while in a sqlite transaction.
* Refactored pull request widget into a base class and two derived classes for a repository pull request and the developer's pull requests.
* Created new Developer pull request widget, "My Pull Requests" which has no configuration and always shows the list of pull requests for all logged in developer Ids. 
* Added step events into cache manager for each account and adds a delay to allow other widgets to query or insert data records between account updates. This makes the first run Query and Repository widgets more responsive. 
* Updated loading page data and template to have a custom loading message that can be changed based on context.
* Updated the Repository widget to use the Repository Name as the widget title if no title is specified by the user.

The My Pull Requests widget requires the repository cache to have been updated before it can find pull requests. It has a new loading message indicating that the search may take several minutes to set expectations. For internal querying this can take upwards of 10 minutes to complete, but we suspect this among the worst-case scenarios and most smaller companies will have much shorter cache sync times.

Once the cache has located the repositories the developer uses then the queries only focus on those repositories.

The My Pull Requests widget does not cause any pull request queries to occur, instead it relies on the existing pull request queries that happen as part of notifications and simply queries the data that is already being populated. It responds to events when that developer pull request is completed to refresh the widget. 

## Validation steps performed

* Lots of manual fresh runs and adding widgets while Caching was happening to see how the widget state handles the caching. Examined logs and debugged the code to ensure the correct behavior was happening.
* Added widget with pre-existing cache and new cache being populated when the widget was created.
* Verification that pull request widget lists same pull requests in Azure Devops.

## PR checklist
